### ISSUE TITLE
support alternative local identifier schemes

### ISSUE THREAD

**Issue by pameyer** [IQSS/dataverse#3654]
To minimize user confusion for repositories migrating to Dataverse, the system used to generate local identifiers should be configurable (local identifier meaning doi:10.5072/FK2/$local_identifier ; this does not involve any changes to external persistent identifier systems).

Currently (~v4.6), this is handled by generating a random alphanumeric sequence and trying to save it (and retrying if consistency errors show that sequence is already in use).  One specific alternative identifier scheme is sequential integers (with the database continuing to server as the source of consistency).  This should allow for one-time (installation? configuration of the start of the sequence as well).

There was some discussion about doing this as an SPI to allow for easier future extensibility.

**Comment by djbrooke** [comment]
Notes from Sprint Planning:

- SPI Y/N - determine
- This will be at the installation level 
- Investigate any relation between this and Handles work
- This will be count and check vs Randomly Generate and check 
- Skipping #s is OK

**Comment by pdurbin** [comment]
I took a swing at this issue, making the smallest possible change in pull request #3710 so I'm moving this issue to Code Review at https://waffle.io/IQSS/dataverse

**Comment by scolapasta** [comment]
That's a good first step and helps define what we want.

Here's what I think:
- The logic for setting the identifier should be in one place -  this PR has it overwriting the current identifier
- the setting should be more generic (not true, false): persistentIdentiferGenerator or something like: we can call the current randomString and the new sequentialNumber; no setting should be needed, by default use randomString
-  then we should create two separate classes for this (this is a start for getting to SPI, even if it would not yet be one, as that work could be work for later if we ever need a third one) for each of the Generators; 
- the RandomStringIdentifierGenerator could just take the existing code
- the SequentialNumberIdentifierGenerator should use some kind of sequence (either in db, or some other way, but not just using the db id)

**Comment by pdurbin** [comment]
I took another swing at this issue in pull request #3713 which I'll move to Code Review in https://waffle.io/IQSS/dataverse

**Comment by scolapasta** [comment]
Reviewed. This is in the right direction.

I have discussed with @michbarsinai and his thoughts on not using a sequence. The main reason is to not have a specific db implementation native query in the code. 

While we do have other native queries in the code, they are more generic SQL for performance issues. I'm sure we do have some postgres specific stuff (e.g. some constraints) but they are not the the java code, and so workarounds can be done in the db. We have a solution for that here too:

We can use a Stored Procedure - this can be defined for the db, but when run from the code, it would use JPA and @NamedStoredProcedureQuery.

As an example:
@NamedStoredProcedureQuery(
	name = "calculate", 
	procedureName = "calculate", 
	parameters = { 
		@StoredProcedureParameter(mode = ParameterMode.IN, type = Double.class, name = "x"), 
		@StoredProcedureParameter(mode = ParameterMode.IN, type = Double.class, name = "y"), 
		@StoredProcedureParameter(mode = ParameterMode.OUT, type = Double.class, name = "sum")
	}
)

and then:
StoredProcedureQuery query = this.em.createNamedStoredProcedureQuery("calculate");
query.setParameter("x", 1.23d);
query.setParameter("y", 4.56d);
query.execute();
Double sum = (Double) query.getOutputParameterValue("sum");

Let's not include the stored procedure in an update script as a) it will be db specific, and b) not everyone will use this, only if they use sequential numbers). Instead we should include in the documentation with the setting (i.e. "if you use this setting, you must add this to your db and here's how:").

Thoughts?

**Comment by scolapasta** [comment]
Also, you didn't do the separation of the logic into separate classes. That's fine, we can do that as part of making this an SPI, if we go that direction with this functionality.

**Comment by pdurbin** [comment]
It looks like that example comes from http://www.thoughts-on-java.org/call-stored-procedures-jpa/ which has more context.

Other resources:

- https://github.com/javaee-samples/javaee7-samples/tree/master/jpa/storedprocedure
- https://blogs.oracle.com/arungupta/entry/jpa_2_1_early_draft

@scolapasta the thing I don't know how to answer is your "you must add this to your db and here's how". Do you know what they must add to their database? I've never used stored procedures before.

**Comment by pdurbin** [comment]
@scolapasta I just read http://www.thoughts-on-java.org/call-stored-procedures-jpa/ and this is the stored procedure the use:

```
CREATE OR REPLACE FUNCTION calculate(
    IN x double precision,
    IN y double precision,
    OUT sum double precision)
  RETURNS double precision AS
$BODY$
BEGIN
    sum = x + y;
END;
$BODY$
LANGUAGE plpgsql
```

This isn't what we want, obviously. If you can write the stored procedure for me, I'm happy to try that approach. Otherwise I recommend we go with pull request #3713 and use a sequence instead.

**Comment by scolapasta** [comment]
Here is how to create stored procedure* in Postgres:

CREATE OR REPLACE FUNCTION generateDatasetIdentifier(
    OUT datasetIdentiier int)
  RETURNS int AS
$BODY$
BEGIN
    select nextval('datasetid_seq') into datasetIdentiier;
END;
$BODY$
  LANGUAGE plpgsql

* this is using 'datasetid_seq' as the name of the sequence. We should probably change this name to use identifier instead of id, since id is often referencing database id's. 

To test and call in SQL, just use: 
select generateDatasetIdentifier();

One more note: with this stored procedure, the sequence needs to be created independently. This is what we need to document for "you must add this to your db and here's how" - how to add the sequence (the create sequence script) and how to add the stored procedure (the above create script).

**Comment by djbrooke** [comment]
We discussed this in sprint planning yesterday. The following items remain:

- Switch this to a stored procedure (see note from @scolapasta above)
- We'll need to add this to the installation guide 
- The old code here was used in a lot of places. Let's spend some time confirming that there will be no undesired effects by making the change here.

**Comment by pdurbin** [comment]
@landreev can you please document :IdentifierGenerationStyle at what will be the future version of http://guides.dataverse.org/en/4.6.1/installation/config.html#database-settings ?

**Comment by pdurbin** [comment]
REST Assured tests were failing after pull request #3713 was merged but after merging pull request #3751 and making a small tweak at 825dbf6 they are passing again:

https://build.hmdc.harvard.edu:8443/job/phoenix.dataverse.org-apitest-develop/

![screen shot 2017-04-05 at 4 11 11 pm](https://cloud.githubusercontent.com/assets/21006/24724900/dcd70628-1a1a-11e7-9671-de9085d6ff8b.png)
